File: m/rls/fiatr.r
file	faatr - file attribute operations
include	rid:rider
include	rid:fidef
include	rid:tidef
include	rid:mxdef
include	<windows.h>

	E_AccFil := "E-Error accessing file [%s]"
	E_SetFil := "E-Error setting file [%s]"
	fiAacc	: * char own = E_AccFil
	fiAset	: * char own = E_SetFil

code	fi_gtm - get file date/time

	diTsta := struct stat

  func	fi_gtm
	spc : * char			; file spec
	val : * tiTval			; the time
	msg : * char			; error message
	()  : int			; fine/fail
  is	nam : [mxLIN] char		;
	han : HANDLE			; file handle
	fil : FILETIME			;
	fi_loc (spc, nam)		; translate the name
	han = CreateFile (nam, GENERIC_READ,
	      FILE_SHARE_READ|FILE_SHARE_WRITE,
	      <>, OPEN_EXISTING, 0, 0)
	if that ne <>			; got a handle
	&& GetFileTime (han, <>,<>,&fil); file time
	&& CloseHandle (han)		; close the file
	&& ti_fnt (&<tiTwin>fil, val)	; convert to our format
	.. fine				;
	fail fi_rep (msg, spc, fiAacc)	; error accessing file
  end 

code	fi_stm - set file date/time

  func	fi_stm
	spc : * char			; file spec
	val : * tiTval			; the time
	msg : * char			; error message
	()  : int			; fine/fail
  is	nam : [mxLIN] char		;
	han : HANDLE			; file handle
	fil : FILETIME			; filetime
	fi_loc (spc, nam)		; translate the name
	ti_tnt (val, &<tiTwin>fil)	; convert to WNT format
	pass fail			;
	han = CreateFile (nam, GENERIC_WRITE,
	      FILE_SHARE_READ|FILE_SHARE_WRITE,
	      <>, OPEN_EXISTING, 0, 0)
	if that ne <>			; yuk
	&& SetFileTime (han, &fil,&fil,&fil); file time
	&& CloseHandle (han)		; close the file
	.. fine				;
	fail fi_rep (msg, spc, fiAacc)	; error accessing file
  end 

code	fi_gat - get attributes

  func	fi_gat
	spc : * char
	atr : * int
	msg : * char			;
	()  : int			;
  is	nam : [mxLIN] char		;
	res : LONG
	fi_loc (spc, nam)		; translate the name
	res = GetFileAttributes (nam)
	if res ne 0xFFFFFFFF		; got the attributes
	   *atr = <int>res if atr	; return attributes
	.. fine				;
	fail fi_rep (msg, spc, fiAacc)	; error accessing file
  end

code	fi_sat - set attributes

  func	fi_sat
	spc : * char
	bic : int			; clear
	bis : int			; set
	msg : * char			;
	()  : int			;
  is	nam : [mxLIN] char		;
	atr : int 			;
	fi_loc (spc, nam)		; translate the name
      repeat				;
	quit if !fi_gat (nam, &atr, <>)	;
	atr &= ~(bic)			; clear some
	atr |= bis			; set some
	SetFileAttributes (nam, <WORD>atr) ;
	pass fine			;
      never				;
	fail fi_rep (msg, spc, fiAset)	; error setting file
  end
end file

code	fi_sta - get local file status

  type	fiTsta
  is	Vdev : word	; dev	device		drive#		handle
	Vnod : word	; ino	inode		0		0
	Vmod : word	; mode
	Vlnk : word	; nlink	links		1		1
	Vuid : word	; user id
	Vgid : word	; group id
	Vred : word	; rdev	redirection	drive#		handle
	Vf00 : word	; dummy
	Vlen : long	; size	byte length
	Vacc : long	; atime	access time
	Vmod : long	; mtime	modify time
	Vcre : long	; ctime	creation time
  end

;	mode bits

	fiMOD_	:= 0xf000
	fiREG	:= 0x8000
	fiBLK	:= 0x6000
	fiNAM	:= 0x5000
	fiDIR	:= 0x4000
	fiCHR	:= 0x2000

	fiMreg(m) := (((m)&(fiMOD_))==fiREG)
	fiMblk(m) := (((m)&(fiMOD_))==fiBLK)
	fiMnam(m) := (((m)&(fiMOD_))==fiNAM)
	fiMdir(m) := (((m)&(fiMOD_))==fiDIR)
	fiMchr(m) := (((m)&(fiMOD_))==fiCHR)

; uid/gid flags

	fiREA_	:= 0x0100 
	fiWRI_	:= 0x0080
	fiEXE_	:= 0x0040

  func	fi_sta
	spc : * char
	han : * FILE
	sta : * fiTsta
  is	loc : struct stat
	res = stat (spc, &stat) if spc
	res = fstat (han, &stat) otherwise
	fail if res eq -1
	sta->Vdev = loc.st_dev
	sta->Vnod = loc.st_ino
	sta->Vmod = loc.st_mode
	sta->Vlnk = loc.st_nlink
	sta->Vuid = loc.st_uid
	sta->Vgid = loc.st_gid
	sta->Vred = loc.st_rdev
	sta->Vlen = loc.st_size
	sta->Vcre = loc.st_ctime
	sta->Vacc = loc.st_atime
	sta->Vmod = loc.st_mtime
  end

struct stat
{	short	st_dev;
	unsigned short	st_ino;
	unsigned short st_mode;
	short	st_nlink;
	unsigned short	st_uid;
	unsigned short	st_gid;
	short	st_rdev;

File: m/rls/rlb/fiatr.c
/* file -  fiatr - file attribute operations */
#include "m:\rid\rider.h"
#include "m:\rid\fidef.h"
#include "m:\rid\tidef.h"
#define E_AccFil  "E-Error accessing file [%s]"
#define E_SetFil  "E-Error setting file [%s]"
static char *fiAacc = E_AccFil;
static char *fiAset = E_SetFil;
#if Dos
#include "m:\rid\dsmod.h"
/* code -  fi_gtm - get file date/time */
int fi_gtm(
char *spc ,
tiTval *tim ,
char *msg )
{ int han ;
  dsTreg reg ;
  dsTseg seg ;
  tiTdos dos ;
  for(;;)  {
    ds_seg (&seg);
    DS = SEG(spc), DX = OFF(spc);
    han = ds_s21 (0x3d00);
    if( CF)break;
    BX=han, ds_r21 (0x5700);
    if( CF){ ds_r21 (0x3e00) ; break;}
    dos.Vdat = DX, dos.Vtim = CX;
    BX=han, ds_r21 (0x3e00);
    ti_fds (&dos, tim);
    return 1;
  } 
   fi_rep (msg, spc, fiAacc);return 0;
} 
/* code -  fi_stm - set file date/time */
int fi_stm(
char *spc ,
tiTval *tim ,
char *msg )
{ tiTdos dos ;
  dsTreg reg ;
  dsTseg seg ;
  int han ;
  int res ;
  ti_tds (tim, &dos);
  for(;;)  {
    ds_seg (&seg);
    DS = SEG(spc), DX = OFF(spc);
    han = ds_s21 (0x3d00);
    if( CF)break;
    CX = dos.Vtim, DX = dos.Vdat;
    BX=han, ds_r21 (0x5701);
    res = !CF;
    BX=han, ds_r21 (0x3e00);
    if( res)return 1;
   break;} 
   fi_rep (msg, spc, fiAset);return 0;
} 
/* code -  fi_gat - get attributes */
int fi_gat(
char *spc ,
int *atr ,
char *msg )
{ dsTreg reg ;
  dsTseg seg ;
  for(;;)  {
    ds_seg (&seg);
    DS = SEG(spc), DX = OFF(spc);
    ds_s21 (0x4300);
    if( CF){ *atr = 0 ; break;}
     *atr = CX;return 1;
   break;} 
   fi_rep (msg, spc, fiAacc);return 0;
} 
/* code -  fi_sat - set attributes */
int fi_sat(
char *spc ,
int bic ,
int bis ,
char *msg )
{ dsTreg reg ;
  dsTseg seg ;
  for(;;)  {
    ds_seg (&seg);
    DS = SEG(spc), DX = OFF(spc);
    ds_s21 (0x4300);
    if( CF)break;
    CX &= ~(bic);
    CX |= bis;
    DS = SEG(spc), DX = OFF(spc);
    ds_s21 (0x4301);
    if( !CF)return 1;
   break;} 
   fi_rep (msg, spc, fiAset);return 0;
} 
#endif 
#if Wnt
/* code -  fi_gtm - get file date/time */
#include "m:\rid\mxdef.h"
#include <windows.h>
#define diTsta  struct stat
int fi_gtm(
char *spc ,
tiTval *val ,
char *msg )
{ char nam [mxLIN];
  HANDLE han ;
  FILETIME fil ;
  fi_loc (spc, nam);
  if (( (han = CreateFile (nam, GENERIC_READ,FILE_SHARE_READ|FILE_SHARE_WRITE,NULL, OPEN_EXISTING, 0, 0)) != NULL)
  &&(GetFileTime (han, NULL,NULL,&fil))
  &&(CloseHandle (han))
  &&(ti_fnt (&(tiTwnt )fil, val))) {
    return 1; }
   fi_rep (msg, spc, fiAacc);return 0;
} 
/* code -  fi_stm - set file date/time */
int fi_stm(
char *spc ,
tiTval *val ,
char *msg )
{ char nam [mxLIN];
  HANDLE han ;
  FILETIME fil ;
  fi_loc (spc, nam);
  ti_tnt (val, &(tiTwnt )fil);
  if (( (han = CreateFile (nam, GENERIC_WRITE,FILE_SHARE_READ|FILE_SHARE_WRITE,NULL, OPEN_EXISTING, 0, 0)) != NULL)
  &&(SetFileTime (han, NULL,NULL,&fil))
  &&(CloseHandle (han))) {
    return 1; }
   fi_rep (msg, spc, fiAacc);return 0;
} 
/* code -  fi_gat - get attributes */
int fi_gat(
char *spc ,
int *atr ,
char *msg )
{ ULONG res ;
  res = GetFileAttributes (spc);
  if ( res != 0xFFFFFFFF) {
     *atr = (int )res;return 1; }
   fi_rep (msg, spc, fiAacc);return 0;
} 
/* code -  fi_sat - set attributes */
int fi_sat(
char *spc ,
int bic ,
int bis ,
char *msg )
{ int atr ;
  for(;;)  {
    if( !fi_gat (spc, &atr, NULL))break;
    atr &= ~(bic);
    atr |= bis;
    if( (SetFileAttributes (spc, (WORD )atr)) != 0)return ( 1 );
   break;} 
   fi_rep (msg, spc, fiAset);return 0;
} 
#endif 
